node/network/bridge/src/lib.rs

```diff
@@ -282,37 +324,211 @@ fn construct_view(live_heads: &[Hash]) -> View {
 	View(live_heads.iter().rev().take(MAX_VIEW_HEADS).cloned().collect())
 }
 
-async fn dispatch_update_to_all(
-	update: NetworkBridgeEvent,
-	event_producers: impl IntoIterator<Item=&fn(NetworkBridgeEvent) -> AllMessages>,
-	ctx: &mut impl SubsystemContext<Message=NetworkBridgeMessage>,
-) -> polkadot_subsystem::SubsystemResult<()> {
-	// collect messages here to avoid the borrow lasting across await boundary.
-	let messages: Vec<_> = event_producers.into_iter()
-		.map(|producer| producer(update.clone()))
-		.collect();
-
-	ctx.send_messages(messages).await
-}
-
 async fn update_view(
-	peers: &HashMap<PeerId, PeerData>,
-	live_heads: &[Hash],
 	net: &mut impl Network,
+	ctx: &mut impl SubsystemContext<Message = NetworkBridgeMessage>,
+	live_heads: &[Hash],
 	local_view: &mut View,
-) -> SubsystemResult<Option<NetworkBridgeEvent>> {
+	validation_peers: &HashMap<PeerId, PeerData>,
+	collation_peers: &HashMap<PeerId, PeerData>,
+) -> SubsystemResult<()> {
 	let new_view = construct_view(live_heads);
-	if *local_view == new_view { return Ok(None) }
+	if *local_view == new_view { return Ok(())  }
+
 	*local_view = new_view.clone();
 
-	let message = WireMessage::ViewUpdate(new_view.clone()).encode();
+	send_validation_message(
+		net,
+		validation_peers.keys().cloned(),
+		WireMessage::ViewUpdate(new_view.clone()),
+	).await?;
+
+	send_collation_message(
+		net,
+		collation_peers.keys().cloned(),
+		WireMessage::ViewUpdate(new_view.clone()),
+	).await?;
+
+	if let Err(e) = dispatch_validation_event_to_all(
+		NetworkBridgeEvent::OurViewChange(new_view.clone()),
+		ctx,
+	).await {
+		log::warn!(target: TARGET, "Aborting - Failure to dispatch messages to overseer");
+		return Err(e)
+	}
+
+	if let Err(e) = dispatch_collation_event_to_all(
+		NetworkBridgeEvent::OurViewChange(new_view.clone()),
+		ctx,
+	).await {
+		log::warn!(target: TARGET, "Aborting - Failure to dispatch messages to overseer");
+		return Err(e)
+	}
+
+	Ok(())
+}
+
+// Handle messages on a specific peer-set. The peer is expected to be connected on that
+// peer-set.
+async fn handle_peer_messages<M>(
+	peer: PeerId,
+	peers: &mut HashMap<PeerId, PeerData>,
+	messages: Vec<WireMessage<M>>,
+	net: &mut impl Network,
+) -> SubsystemResult<Vec<NetworkBridgeEvent<M>>> {
+	let peer_data = match peers.get_mut(&peer) {
+		None => {
+			net.report_peer(peer, UNCONNECTED_PEERSET_COST).await?;
+
+			return Ok(Vec::new());
+		},
+		Some(d) => d,
+	};
+
+	let mut outgoing_messages = Vec::with_capacity(messages.len());
+	for message in messages {
+		outgoing_messages.push(match message {
```

Comment:
This loop might be cleaner as

```rust
messages.iter().filter_map(|message| ...).collect()
```

Reply:
another thing I tried at first :) 

The problem is the `.await` statements within the loop